llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp

```diff
@@ -2796,6 +2796,25 @@ Value *LibCallSimplifier::optimizeSqrt(CallInst *CI, IRBuilderBase &B) {
   return copyFlags(*CI, FabsCall);
 }
 
+Value *LibCallSimplifier::optimizeFMod(CallInst *CI, IRBuilderBase &B) {
+  SimplifyQuery SQ(DL, TLI, DT, AC, CI, true, true, DC);
+
+  // fmod(x,y) can set errno if y == 0 or x == +/-inf.
+  KnownFPClass Known0 = computeKnownFPClass(CI->getOperand(0), fcInf,
+                                            /*Depth=*/0, SQ);
+  if (Known0.isKnownNeverInfinity()) {
+    KnownFPClass Known1 = computeKnownFPClass(CI->getOperand(1), fcZero,
+                                              /*Depth=*/0, SQ);
+    if (Known1.isKnownNeverZero()) {
```

Comment:
This is wrong if DAZ is enabled, this needs to account of the denormal mode. Can you add a negative test with "denormal-fp-math"="preserve-sign,preserve-sign" and/or ="dynamic,dynamic"? This should be checking isKnownNeverLogicalZero

Reply:
Thanks - my testing (https://gist.github.com/davemgreen/f6b084618c24db2b0996cc618cd8ca05) was on a normal linux system. There is a test in the end of ad3ad15229cc65874eae173137e47b5c817d77ea, and I've changed it to isKnownNeverLogicalZero.